File: core/DecodeString/DecodeString.py
def decodeString(s: str) -> str:
    stack = []
    for ch in s:
        if ch == "]":
            sub_str = ""
            while stack and stack[-1] != "[":
                sub_str = stack.pop() + sub_str
            if stack:
                stack.pop()
                num = ""
                while stack and stack[-1].isdigit():
                    num = stack.pop() + num
                num = int(num)
                temp = ""
                for _ in range(num):
                    temp += sub_str
                stack.append(temp)
        else:
            stack.append(ch)

    return "".join(stack)

if __name__ == '__main__':
    print(decodeString("100[leetcode]"))

File: core/DecodeString/DecodeString.java
package com.dbc;

import java.util.Stack;

public class DecodeString {
    public String decodeString(String s) {
        Stack<String> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == ']') {
                StringBuilder subStr = new StringBuilder();
                while (!stack.isEmpty() && !stack.peek().equals("[")) subStr.insert(0, stack.pop());
                if (!stack.isEmpty()) {
                    stack.pop();
                    StringBuilder numStr = new StringBuilder();
                    while (!stack.isEmpty() && stack.peek().length() == 1 && Character.isDigit(stack.peek().charAt(0))) numStr.insert(0, stack.pop());
                    int num = Integer.parseInt(numStr.toString());
                    for (int j = 0; j < num; j++) stack.add(subStr.toString());
                }
            } else {
                stack.add(Character.toString(s.charAt(i)));
            }
        }
        StringBuilder res = new StringBuilder();
        while (!stack.isEmpty()) {
            res.insert(0, stack.pop());
        }
        return res.toString();
    }
}
